drag_and_drop_v2/drag_and_drop_v2.py

```diff
@@ -294,101 +294,71 @@ def studio_submit(self, submissions, suffix=''):
+        elif self.mode == self.STANDARD_MODE:
+            return self._drop_item_standard(item_attempt)
         else:
-            zone = self._get_zone_by_uid(attempt['zone'])
-        if not zone:
-            raise JsonHandlerError(400, "Item zone data invalid.")
+            raise JsonHandlerError(500, _("Unknown DnDv2 mode {mode} - course is misconfigured").format(self.mode))
 
-        if self._is_finished():
-            overall_feedback = self.data['feedback']['finish']
+    @XBlock.json_handler
+    def do_attempt(self, data, suffix=''):
+        self._validate_attempt()
 
-        # don't publish the grade if the student has already completed the problem
-        if not self.completed:
-            if self._is_finished():
-                self.completed = True
-            try:
-                self.runtime.publish(self, 'grade', {
-                    'value': self._get_grade(),
-                    'max_value': self.weight,
-                })
-            except NotImplementedError:
-                # Note, this publish method is unimplemented in Studio runtimes,
-                # so we have to figure that we're running in Studio for now
-                pass
+        self.num_attempts += 1
+        self._mark_complete_and_publish_grade()
 
-        self.runtime.publish(self, 'edx.drag_and_drop_v2.item.dropped', {
-            'item_id': item['id'],
-            'location': zone.get("title"),
-            'location_id': zone.get("uid"),
-            'is_correct': is_correct,
-        })
+        __, placed, correct = self._get_item_raw_stats()
+        placed_ids, correct_ids = set(placed), set(correct)
+        misplaced_ids = placed_ids - correct_ids
 
-        if self.mode == self.ASSESSMENT_MODE:
-            # In assessment mode we don't send any feedback on drop.
-            result = {}
+        feedback_msgs = [_('Correctly placed {correct_count} items, misplaced {misplaced_count} items.').format(
+            correct_count=len(correct_ids),
+            misplaced_count=len(misplaced_ids)
+        )]
+
+        if misplaced_ids:
+            feedback_msgs.append(_('Misplaced items were returned to item bank.'))
         else:
-            result = {
-                'correct': is_correct,
-                'finished': self._is_finished(),
-                'overall_feedback': overall_feedback,
-                'feedback': feedback
-            }
+            feedback_msgs.append(self.data['feedback']['finish'])
 
-        return result
+        for item_id in misplaced_ids:
+            del self.item_state[item_id]
+
+        if not self.attemps_remain:
+            feedback_msgs.append(_('Final attempt was used, final score is {score}').format(score=self._get_grade()))
+
+        return {
+            'num_attempts': self.num_attempts,
+            'misplaced_items': list(misplaced_ids),
+            'feedback': '\n'.join(feedback_msgs),
+            'attempts_remain': self.attemps_remain
```

Comment:
Nit: `attempts_remain` is somewhat redundant since we send `num_attempts` and the client knows the value of `max_attempts`.